src/modules-load/modules-load.c

```diff
@@ -39,45 +88,249 @@ static int parse_proc_cmdline_item(const char *key, const char *value, void *dat
+        if (bytes < 0)
+                return log_error_errno(errno, "Failed to send '%s' to thread pool: %m", module);
+
+        return 0;
+}
+
+static int dequeue_module_to_load(int sock, char *buffer, size_t buffer_len) {
+        ssize_t bytes;
+
+        assert(sock >= 0);
+        assert(buffer);
+
+        /* Dequeue one module to be loaded from the socket pair. In case no more
+         * modules are present, recv() will return 0. */
+        bytes = recv(sock, buffer, buffer_len, 0);
+        if (bytes == 0)
+                return 0;
+        if (bytes < 0)
+                return negative_errno();
+        if ((size_t)bytes == buffer_len)
+                return log_error_errno(SYNTHETIC_ERRNO(E2BIG), "Dequeued module name too long");
+
+        buffer[bytes] = '\0';
+
+        return 1;
+}
+
+static int run_prober(int sock) {
+        _cleanup_(sym_kmod_unrefp) struct kmod_ctx *ctx = NULL;
+        char buffer[MODULE_NAME_MAX_LEN + 1];
+        int ret = 0, r;
+
+        r = module_setup_context(&ctx);
+        if (r < 0)
+                return log_error_errno(r, "Failed to initialize libkmod context: %m");
+
         for (;;) {
-                _cleanup_free_ char *line = NULL;
-                int k;
+                r = dequeue_module_to_load(sock, buffer, sizeof(buffer));
+                if (ERRNO_IS_NEG_TRANSIENT(r))
+                        continue;
 
-                k = read_stripped_line(f, LONG_LINE_MAX, &line);
-                if (k < 0)
-                        return log_error_errno(k, "Failed to read file '%s': %m", pp);
-                if (k == 0)
+                if (r < 0)
+                        return log_error_errno(r, "Failed to receive from module queue: %m");
+
+                if (r == 0) {
+                        log_debug("No more queued modules, terminate thread");
                         break;
+                }
 
-                if (isempty(line))
-                        continue;
-                if (strchr(COMMENTS, *line))
-                        continue;
+                r = module_load_and_warn(ctx, buffer, /* verbose = */ true);
+                if (r != -ENOENT)
+                        RET_GATHER(ret, r);
+        }
 
-                k = module_load_and_warn(ctx, line, true);
-                if (k == -ENOENT)
-                        continue;
-                RET_GATHER(r, k);
+        return ret;
+}
+
+static void *prober_thread(void *arg) {
+        int sock = PTR_TO_FD(arg);
+        return INT_TO_PTR(run_prober(sock));
+}
+
+static int create_worker_threads(
+                size_t n_threads,
+                void *arg,
+                pthread_t **ret_threads,
+                size_t *ret_n_threads) {
+        _cleanup_free_ pthread_t *new_threads = NULL;
+        size_t created_threads;
+        sigset_t ss, saved_ss;
+        int r;
+
+        assert(ret_threads);
+        assert(ret_n_threads);
+
+        if (n_threads == 0) {
+                *ret_n_threads = 0;
```

Comment:
please also initialize `*ret_threads = NULL`. We have this rule that return parameters should be fully initialized on success, and not touched at all in error cases. This is a success case, hence you should initialize it.